### PR TITLE
Rewrite run_parser* functions to always return a mod_ty

### LINKED ISSUE #15: Rewrite of run_parser* functions
They should all return a `mod_ty`. The caller can then use it to create the necessary `PyObject *`.

### PR DESCRIPTION
**Pull request by lysnikolaou**
Closes #15.

### CODE REVIEW

File: Parser/pegen/peg_api.c [hunk 1]
```diff
@@ -4,25 +4,35 @@
...
-PyPegen_ASTFromFile(const char *filename, PyArena *arena)
+PyPegen_ASTFromFile(const char *filename, PyObject *filename_ob, PyArena *arena)
 {
-    return run_parser_from_file(filename, parse_start, RAW_AST_OBJECT, arena);
+    return run_parser_from_file(filename, parse_start, filename_ob, arena);
 }
 
 PyCodeObject *
-PyPegen_CodeObjectFromString(const char *str, PyArena *arena)
+PyPegen_CodeObjectFromString(const char *str, PyObject *filename_ob, PyArena *arena)
```
Comment (gvanrossum) [inline Parser/pegen/peg_api.c]: Why does this need the caller to create and free the arena? Once we're done generating the code we don't need it any more. (This is different for the previous two.)

File: Parser/pegen/peg_api.c [hunk 2]
```diff
@@ -4,25 +4,35 @@
 #include "pegen.h"
 
 mod_ty
-PyPegen_ASTFromString(const char *str, PyArena *arena)
+PyPegen_ASTFromString(const char *str, PyObject *filename_ob, PyArena *arena)
```
Comment (gvanrossum) [inline Parser/pegen/peg_api.c]: Remind me, what's the ownership issue that requires the caller to pass the filename? Can't this function make the `PyUnicode_FromString("<string>")` call? (And ditto for the next function.)

File: Parser/pegen/peg_api.c [hunk 3]
```diff
@@ -4,25 +4,35 @@
...
-    return run_parser_from_file(filename, parse_start, RAW_AST_OBJECT, arena);
+    return run_parser_from_file(filename, parse_start, filename_ob, arena);
 }
 
 PyCodeObject *
-PyPegen_CodeObjectFromString(const char *str, PyArena *arena)
+PyPegen_CodeObjectFromString(const char *str, PyObject *filename_ob, PyArena *arena)
 {
-    return run_parser_from_string(str, parse_start, CODE_OBJECT, arena);
+    mod_ty res = run_parser_from_string(str, parse_start, filename_ob, arena);
```
Comment (gvanrossum) [inline Parser/pegen/peg_api.c]: Shouldn't this call `PyPegen_ASTFromString()`? That would make the relationship between the two functions clearer. The compiler can auto-inline it if it thinks it'll matter. (And ditto for the next function.)

File: Parser/pegen/pegen.c [hunk 4]
```diff
@@ -559,89 +559,65 @@ run_parser(Parser *p, int mode)
...
+mod_ty
+run_parser_from_string(const char *str, mod_ty(*parse_func)(Parser *), PyObject *filename_ob, PyArena *arena)
 {
-    void *result = NULL;
     struct tok_state *tok = PyTokenizer_FromString(str, 1);
     if (tok == NULL) {
         return NULL;
     }
+    tok->filename = filename_ob;
+    Py_INCREF(filename_ob);
```
Comment (gvanrossum) [inline Parser/pegen/pegen.c]: This warrants a comment explaining that the tokenizer now owns it. (I had to look it up in `PyTokenizer_Free()` to be sure. :-)

File: Python/pythonrun.c [hunk 5]
```diff
@@ -1036,7 +1036,7 @@ PyRun_StringFlags(const char *str, int start, PyObject *globals,
     // I left this here so when you run 'nm python | grep PyPegen' you can see
     // that the symbol is included and this function is called. Obviously this is
     // wrong, and the function needs to be changed to return
-    mod = PyPegen_ASTFromString(str, arena);
+    mod = PyPegen_ASTFromString(str, filename, arena);
```
[the remark this answers is not in the document]
Comment (lysnikolaou) [reply Python/pythonrun.c]: This comment was added by @pablogsal.

Note that this is changed in #19.

**Comment by lysnikolaou** [comment]
> I still think we should switch EXTENSION_SUFFIX to be much smaller and move most of that code to a separate file.

I have this almost ready and was thinking of opening a new PR for it. What do you think?

File: Parser/pegen/peg_api.c (hunk 1 above)
Comment (lysnikolaou) [reply Parser/pegen/peg_api.c]: No need to, I just did this like that in order to be consistent with `PyAST_CompileObject` which expects the arena as an argument.

File: Python/pythonrun.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (pablogsal) [reply Python/pythonrun.c]: > Re-reading the comment, does this still crash? I don't understand the utility of the 'nm' thing you mention.

I left this in my original commit to checking that the symbols were being included in `libpython` to check for progress (the symbols were there and exported correctly). The reason is that if you don't call anything the linker does not include them so I needed to do a dummy call.

File: Parser/pegen/peg_api.c (hunk 2 above)
Comment (lysnikolaou) [reply Parser/pegen/peg_api.c]: Once more, I just did this like this just to be consistent with `PyParser_ASTFromStringObject`.

Comment (gvanrossum) [reply Parser/pegen/peg_api.c]: I think the situation is a little different -- those APIs are constrained by the need to first create the Concrete Syntax Tree and then convert that to the parse tree. Let's do the simplest thing that works for our use cases for now.

Comment (lysnikolaou) [reply Parser/pegen/peg_api.c]: Alright!

**Comment by lysnikolaou** [comment]
> Better, but I'd still like to see if we can easily drop the `filename_ob` parameters.

Just pushed a fix for this.

**Comment by lysnikolaou** [comment]
> Bang!

Yes!

**Pull request merged**